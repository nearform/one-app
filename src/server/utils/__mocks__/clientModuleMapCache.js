/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const clientModuleMapCache = jest.requireActual('../clientModuleMapCache');

clientModuleMapCache.setClientModuleMapCache({
  key: '123',
  modules: {
    'test-root': {
      node: {
        url: 'https://example.com/cdn/test-root/2.2.2/test-root.node.js',
        integrity: '4y45hr',
      },
      browser: {
        url: 'https://example.com/cdn/test-root/2.2.2/test-root.browser.js',
        integrity: 'dhhfsdfwer',
      },
      legacyBrowser: {
        url: 'https://example.com/cdn/test-root/2.2.2/test-root.legacy.browser.js',
        integrity: '567ty',
      },
    },
    a: {
      node: {
        url: 'https://example.com/cdn/a/2.2.2/a.node.js',
        integrity: 'awefs',
      },
      browser: {
        url: 'https://example.com/cdn/a/2.2.2/a.browser.js',
        integrity: 'fhgnt543',
      },
      legacyBrowser: {
        url: 'https://example.com/cdn/a/2.2.2/a.legacy.browser.js',
        integrity: '7567ee',
      },
    },
    c: {
      node: {
        url: 'https://example.com/cdn/c/2.2.2/c.node.js',
        integrity: '3535eqr',
      },
      browser: {
        url: 'https://example.com/cdn/c/2.2.2/c.browser.js',
        integrity: '323egdsbf',
      },
      legacyBrowser: {
        url: 'https://example.com/cdn/c/2.2.2/c.legacy.browser.js',
        integrity: 'd8vdfdfv',
      },
    },
    b: {
      node: {
        url: 'https://example.com/cdn/b/2.2.2/b.node.js',
        integrity: '4y45hr',
      },
      browser: {
        url: 'https://example.com/cdn/b/2.2.2/b.browser.js',
        integrity: 'yhrtrhw3',
      },
      legacyBrowser: {
        url: 'https://example.com/cdn/b/2.2.2/b.legacy.browser.js',
        integrity: '7567ee',
      },
    },
  },
});

export const getClientModuleMapCache = jest.fn(clientModuleMapCache.getClientModuleMapCache);
export const setClientModuleMapCache = jest.fn(clientModuleMapCache.setClientModuleMapCache);
