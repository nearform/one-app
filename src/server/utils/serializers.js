import { Map as iMap } from 'immutable';

import transit from '../../universal/utils/transit';

// eslint-disable-next-line import/prefer-default-export
export function serializeClientInitialState(clientInitialState) {
  // try to build the full state, this _might_ fail (ex: 'Error serializing unrecognized object')
  try {
    return transit.toJSON(clientInitialState);
  } catch (err) {
    console.error('encountered an error serializing full client initial state', err);

    // clear out an internal cache that corrupts the serialization generated on the next call
    // TODO: understand transit-js and transit-immutable-js internals to properly fix the bug
    // for now stop the bleeding
    transit.toJSON('clear out an internal cache');
  }

  // can't send all the work we've done to build the state, but we can still give the app what it
  // needs to start in the browser
  // this _shouldn't_ ever throw, but just in case...
  try {
    return transit.toJSON(
      iMap({
        config: clientInitialState.get('config'),
        holocron: clientInitialState.get('holocron'),
      })
    );
  } catch (err) {
    transit.toJSON('clear out an internal cache, again');
    // something is really wrong
    console.error('unable to build the most basic initial state for a client to startup', err);
    throw err;
  }
}
