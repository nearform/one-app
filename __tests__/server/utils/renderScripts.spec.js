/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { fromJS, Set as iSet } from 'immutable';
import uuid from 'uuid/v4';

import {
  getModules,
  renderScripts,
  renderScriptTag,
  renderI18nScript,
  renderChunkAssetScripts,
  renderModuleScripts,
  renderInitialStateScriptTag,
} from '../../../src/server/utils/renderScripts';
import { serializeClientInitialState } from '../../../src/server/utils/serializers';
import { getClientModuleMapCache } from '../../../src/server/utils/clientModuleMapCache';

jest.mock('uuid/v4');
jest.mock('../../../src/server/utils/readJsonFile');
jest.mock('../../../src/server/utils/clientModuleMapCache');

jest.spyOn(console, 'error').mockImplementation(() => null);

const rootModuleName = 'test-root';
const activeLocale = 'en-US';
const publicPath = '/_/static';

describe('renderScripts', () => {
  const clientInitialState = fromJS({
    intl: { activeLocale },
    config: { rootModuleName },
    holocron: { loaded: iSet([rootModuleName]) },
  });

  const loadedState = fromJS({
    intl: { activeLocale },
    config: { rootModuleName },
    holocron: { loaded: iSet([]) },
  });

  it('should fail without initial state', () => {
    expect.assertions(1);
    expect(() => renderScripts()).toThrow();
  });

  it('should render all app scripts with bare state as default', () => {
    expect.assertions(15);

    const scripts = renderScripts({ clientInitialState });

    const split = scripts.split('\n');
    const endsWithAppJs = split[split.length - 1].startsWith('<script src="/_/static/app.js"');
    const startsWithInitialState = scripts.startsWith(
      '<script id="initial-state" nonce="'
    );

    expect(startsWithInitialState).toBe(true);
    expect(endsWithAppJs).toBe(true);
    // check for default initial state keys
    expect(scripts).toContain("window.__render_mode__ = 'hydrate';");
    expect(scripts).toContain('window.__webpack_public_path__ = "/_/static/";');
    expect(scripts).toContain("window.__holocron_module_bundle_type__ = 'browser';");
    expect(scripts).toContain('window.__CLIENT_HOLOCRON_MODULE_MAP__ = {"key":"123",');
    expect(scripts).toContain('window.__INITIAL_STATE__ = "');
    // ensure expected script tags
    expect(scripts).toContain('<script src="/_/static/i18n/en-US.js"');
    expect(scripts).toContain('<script src="/_/static/app~vendors.js"');
    expect(scripts).toContain('<script src="/_/static/vendors.js"');
    expect(scripts).toContain('<script src="/_/static/runtime.js"');
    expect(scripts).toContain(
      '<script src="https://example.com/cdn/test-root/2.2.2/test-root.browser.js?key=123" crossorigin="anonymous" integrity="dhhfsdfwer"></script>'
    );
    // defaults are called
    expect(uuid).toHaveBeenCalled();
    expect(getClientModuleMapCache).toHaveBeenCalled();
    expect(scripts).toMatchSnapshot();
  });

  it('should render scripts in devMode', () => {
    expect.assertions(3);

    const scripts = renderScripts({
      clientInitialState,
      devMode: true,
    });

    expect(scripts).not.toContain('integrity="');
    expect(scripts).toContain("window.__render_mode__ = 'hydrate';");
    expect(scripts).toContain('window.__webpack_public_path__ = "/_/static/";');
  });

  it('should render scripts in static mode', () => {
    expect.assertions(3);

    const scripts = renderScripts({
      clientInitialState,
      devMode: true,
      isStatic: true,
    });

    expect(scripts).not.toContain('integrity="');
    expect(scripts).toContain("window.__render_mode__ = 'render';");
    expect(scripts).toContain('window.__webpack_public_path__ = "/_/static/";');
  });

  it('should render scripts in legacy mode', () => {
    expect.assertions(3);

    const scripts = renderScripts({
      clientInitialState,
      legacy: true,
      crossOrigin: null,
      scriptNonce: null,
    });

    expect(scripts).toContain('integrity="');
    expect(scripts).toContain("window.__render_mode__ = 'hydrate';");
    expect(scripts).toContain('window.__webpack_public_path__ = "/_/static/";');
  });

  it('should render scripts in legacy and dev mode, with custom publicPath', () => {
    expect.assertions(3);

    const scripts = renderScripts({
      clientInitialState,
      devMode: true,
      legacy: true,
      crossOrigin: null,
      scriptNonce: null,
      publicPath: '/static/',
    });

    expect(scripts).not.toContain('integrity="');
    expect(scripts).toContain("window.__render_mode__ = 'hydrate';");
    expect(scripts).toContain('window.__webpack_public_path__ = "/static/";');
  });

  it('should render scripts without loaded modules', () => {
    expect.assertions(2);

    const scripts = renderScripts({
      clientInitialState: loadedState,
      devMode: true,
      crossOrigin: null,
      scriptNonce: null,
    });

    expect(scripts).toContain("window.__render_mode__ = 'hydrate';");
    expect(scripts).not.toContain(
      '<script src="https://example.com/cdn/test-root/2.2.2/test-root.browser.js?key=123" crossorigin="anonymous" integrity="dhhfsdfwer"></script>'
    );
  });

  it('should render scripts without defaults', () => {
    expect.assertions(3);

    const scripts = renderScripts({
      clientInitialState: loadedState,
      devMode: true,
      crossOrigin: null,
      scriptNonce: null,
      publicPath: null,
      clientModuleMapCache: null,
    });

    expect(scripts).toMatchSnapshot();
    expect(scripts).toContain("window.__render_mode__ = 'hydrate';");
    expect(scripts).not.toContain(
      '<script src="https://example.com/cdn/test-root/2.2.2/test-root.browser.js?key=123" crossorigin="anonymous" integrity="dhhfsdfwer"></script>'
    );
  });
});

describe('getModules', () => {
  it('should prep modules for consumption', () => {
    const bundle = 'browser';
    const modules = iSet(['a', 'b', 'test-root']);
    const moduleMap = getClientModuleMapCache();
    const set = getModules(
      modules,
      moduleMap[bundle],
      rootModuleName
    );
    const [{ [bundle]: { url } }] = set[0];

    expect(set).toMatchSnapshot();
    expect(url).toContain(rootModuleName);
    expect(url).toContain(bundle);
  });

  it('should load legacy modules', () => {
    const bundle = 'legacyBrowser';
    const modules = iSet(['a', 'b', 'test-root']);
    const moduleMap = getClientModuleMapCache();
    const set = getModules(
      modules,
      moduleMap[bundle],
      rootModuleName
    );
    const [{ [bundle]: { url } }] = set[0];

    expect(set).toMatchSnapshot();
    expect(url).toContain(rootModuleName);
    expect(url).toContain('legacy');
  });
});

describe('renderModuleScripts', () => {
  const bundle = 'browser';
  const crossOrigin = 'anonymous';
  const moduleMap = getClientModuleMapCache();
  const clientInitialState = fromJS({
    intl: { activeLocale },
    config: { rootModuleName },
    holocron: { loaded: iSet([rootModuleName]) },
  });

  it('adds cache busting key from module map to each module script src if NODE_ENV is production', () => {
    expect.assertions(3);

    const moduleScripts = renderModuleScripts({
      clientInitialState,
      moduleMap,
      bundle,
      crossOrigin,
      devMode: false,
    });
    const expected = '<script src="https://example.com/cdn/test-root/2.2.2/test-root.browser.js?key=123" crossorigin="anonymous" integrity="dhhfsdfwer"></script>';

    expect(moduleScripts).toContain(expected);
    expect(moduleScripts.startsWith(expected)).toBe(true);
    expect(moduleScripts.endsWith(expected)).toBe(true);
  });

  it('does not add cache busting key from module map to each module script src if NODE_ENV is development', () => {
    expect.assertions(3);

    const moduleScripts = renderModuleScripts({
      clientInitialState,
      moduleMap,
      bundle,
      crossOrigin,
      devMode: true,
    });
    const expected = '<script src="https://example.com/cdn/test-root/2.2.2/test-root.browser.js" crossorigin="anonymous"></script>';

    expect(moduleScripts).toContain(expected);
    expect(moduleScripts.startsWith(expected)).toBe(true);
    expect(moduleScripts.endsWith(expected)).toBe(true);
  });

  it('sends a rendered page with cross origin scripts', () => {
    expect.assertions(3);

    const moduleScripts = renderModuleScripts({
      clientInitialState,
      moduleMap,
      bundle,
      devMode: false,
      crossOrigin,
    });
    const expected = '<script src="https://example.com/cdn/test-root/2.2.2/test-root.browser.js?key=123" crossorigin="anonymous" integrity="dhhfsdfwer"></script>';

    expect(moduleScripts).toContain(expected);
    expect(moduleScripts.startsWith(expected)).toBe(true);
    expect(moduleScripts.endsWith(expected)).toBe(true);
  });

  it('send a rendered page with correctly ordered modules', () => {
    expect.assertions(1);

    const moduleScripts = renderModuleScripts({
      clientInitialState: clientInitialState.mergeDeep({
        holocron: fromJS({
          loaded: iSet(['a', 'b', 'test-root', 'c']),
        }),
      }),
      moduleMap,
      bundle,
      devMode: true,
    });
    const expected = [
      '<script src="https://example.com/cdn/test-root/2.2.2/test-root.browser.js">',
      '<script src="https://example.com/cdn/a/2.2.2/a.browser.js">',
      '<script src="https://example.com/cdn/b/2.2.2/b.browser.js">',
      '<script src="https://example.com/cdn/c/2.2.2/c.browser.js">',
    ];

    expect(
      moduleScripts
        .split('\n')
        .map((str) => str.replace('</script>', ''))
    ).toEqual(expected);
  });

  it('send a rendered page keeping correctly ordered modules if NODE_ENV is production', () => {
    expect.assertions(1);

    const moduleScripts = renderModuleScripts({
      clientInitialState: clientInitialState.mergeDeep({
        holocron: fromJS({
          loaded: iSet(['test-root', 'a', 'b', 'c']),
        }),
      }),
      moduleMap,
      bundle,
      devMode: false,
    });
    const expected = [
      '<script src="https://example.com/cdn/test-root/2.2.2/test-root.browser.js?key=123"  integrity="dhhfsdfwer">',
      '<script src="https://example.com/cdn/a/2.2.2/a.browser.js?key=123"  integrity="fhgnt543">',
      '<script src="https://example.com/cdn/b/2.2.2/b.browser.js?key=123"  integrity="yhrtrhw3">',
      '<script src="https://example.com/cdn/c/2.2.2/c.browser.js?key=123"  integrity="323egdsbf">',
    ];

    expect(moduleScripts.split('\n').map((str) => str.replace('</script>', ''))).toEqual(expected);
  });

  it('send a rendered page with module script tags with integrity attribute if NODE_ENV is production', () => {
    expect.assertions(1);

    const moduleScripts = renderModuleScripts({
      clientInitialState: clientInitialState.mergeDeep({
        holocron: fromJS({
          loaded: iSet([rootModuleName]),
        }),
      }),
      moduleMap,
      bundle,
      devMode: false,
      crossOrigin,
    });
    const expected = '<script src="https://example.com/cdn/test-root/2.2.2/test-root.browser.js?key=123" crossorigin="anonymous" integrity="dhhfsdfwer"></script>';

    expect(moduleScripts).toContain(
      expected
    );
  });
});

describe('renderI18nScript', () => {
  const crossOrigin = 'anonymous';

  it('should render activeLocale', () => {
    const clientInitialState = fromJS({
      intl: { activeLocale },
    });
    const i18nScript = renderI18nScript({
      clientInitialState,
      crossOrigin,
      publicPath,
    });

    expect(i18nScript).toContain(
      `<script src="/_/static/i18n/en-US.js" crossorigin="${crossOrigin}"></script>`
    );
  });

  it('should render activeLocale without crossorigin', () => {
    const clientInitialState = fromJS({
      intl: { activeLocale: 'es-ES' },
    });

    const i18nScript = renderI18nScript({
      clientInitialState,
      publicPath,
    });

    expect(i18nScript).toContain(
      '<script src="/_/static/i18n/es-ES.js"></script>'
    );
  });

  it('should return empty string if i18n file not found', () => {
    const clientInitialState = fromJS({
      intl: { activeLocale: '1zx' },
    });

    const i18nScript = renderI18nScript({
      clientInitialState,
      publicPath,
    });

    expect(i18nScript).toEqual('');
  });
});

describe('renderChunkAssetScripts', () => {
  const legacy = false;

  it('send a rendered page with correctly ordered modules', () => {
    expect.assertions(1);

    const chunkAssets = renderChunkAssetScripts({
      legacy,
      publicPath,
      integrity: false,
    });
    const expected = [
      `<script src="${publicPath}/app~vendors.js">`,
      `<script src="${publicPath}/runtime.js">`,
      `<script src="${publicPath}/vendors.js">`,
    ];

    expect(chunkAssets.split('\n').map((str) => str.replace('</script>', ''))).toEqual(expected);
  });

  it('send a rendered page keeping correctly ordered modules if NODE_ENV is production', () => {
    expect.assertions(3);

    const currentPath = `${publicPath}/legacy`;
    const chunkAssets = renderChunkAssetScripts({
      legacy,
      publicPath: currentPath,
      integrity: true,
    });

    expect(chunkAssets).toContain(`<script src="${currentPath}/app~vendors.js" integrity="`);
    expect(chunkAssets).toContain(`<script src="${currentPath}/runtime.js" integrity="`);
    expect(chunkAssets).toContain(`<script src="${currentPath}/vendors.js" integrity="`);
  });
});

describe('renderInitialStateScriptTag', () => {
  const bundle = 'browser';
  const webpackPath = JSON.stringify('/');
  const moduleMap = JSON.stringify(getClientModuleMapCache());
  const initialState = serializeClientInitialState(fromJS({
    intl: { activeLocale },
    config: { rootModuleName },
    holocron: { loaded: iSet([rootModuleName]) },
  }));

  it('should render initial state inside a script tag body', () => {
    expect.assertions(8);
    const id = 'initial-state';
    const scriptNonce = uuid();
    const renderedState = renderInitialStateScriptTag({
      id,
      publicPath: webpackPath,
      bundle,
      moduleMap,
      initialState,
      scriptNonce,
    });
    expect(renderedState).toMatchSnapshot();
    expect(renderedState).toContain(`<script id="${id}" nonce="${scriptNonce}">`);
    expect(renderedState).toContain("window.__render_mode__ = 'hydrate';");
    expect(renderedState).toContain(`window.__webpack_public_path__ = ${webpackPath};`);
    expect(renderedState).toContain(`window.__holocron_module_bundle_type__ = '${bundle}';`);
    expect(renderedState).toContain(`window.__CLIENT_HOLOCRON_MODULE_MAP__ = ${moduleMap}`);
    expect(renderedState).toContain(`window.__INITIAL_STATE__ = ${initialState};`);
    expect(renderedState).toContain('</script>');
  });
});

describe('renderScriptTag', () => {
  const src = '/app.js';
  const attrs = ['crossorigin="anonymous"'];
  const body = 'console.log(this);';

  it('should render bare bones script tag', () => {
    expect.assertions(2);
    expect(renderScriptTag()).toEqual('<script></script>');
    expect(renderScriptTag()).toMatchSnapshot();
  });

  it('should render script tag with src', () => {
    expect.assertions(2);
    expect(renderScriptTag(src)).toEqual(`<script src="${src}"></script>`);
    expect(renderScriptTag(src)).toMatchSnapshot();
  });

  it('should render script tag with src and attrs', () => {
    expect.assertions(2);
    expect(renderScriptTag(src, [...attrs])).toEqual(
      `<script src="${src}" crossorigin="anonymous"></script>`
    );
    expect(renderScriptTag(src, [...attrs])).toMatchSnapshot();
  });

  it('should render script tag with src, attrs and body', () => {
    expect.assertions(2);
    expect(renderScriptTag(src, [...attrs], body)).toEqual(
      `<script src="${src}" crossorigin="anonymous">\n${body}\n</script>`
    );
    expect(renderScriptTag(src, [...attrs], body)).toMatchSnapshot();
  });
});
