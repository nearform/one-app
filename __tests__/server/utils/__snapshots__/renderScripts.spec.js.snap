// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getModules should load legacy modules 1`] = `
Array [
  Array [
    Object {
      "baseUrl": "https://example.com/cdn/test-root/2.2.2/",
      "legacyBrowser": Object {
        "integrity": "567ty",
        "url": "https://example.com/cdn/test-root/2.2.2/test-root.legacy.browser.js",
      },
    },
    "test-root",
  ],
  Array [
    Object {
      "baseUrl": "https://example.com/cdn/a/2.2.2/",
      "legacyBrowser": Object {
        "integrity": "7567ee",
        "url": "https://example.com/cdn/a/2.2.2/a.legacy.browser.js",
      },
    },
    "a",
  ],
  Array [
    Object {
      "baseUrl": "https://example.com/cdn/b/2.2.2/",
      "legacyBrowser": Object {
        "integrity": "7567ee",
        "url": "https://example.com/cdn/b/2.2.2/b.legacy.browser.js",
      },
    },
    "b",
  ],
]
`;

exports[`getModules should prep modules for consumption 1`] = `
Array [
  Array [
    Object {
      "baseUrl": "https://example.com/cdn/test-root/2.2.2/",
      "browser": Object {
        "integrity": "dhhfsdfwer",
        "url": "https://example.com/cdn/test-root/2.2.2/test-root.browser.js",
      },
    },
    "test-root",
  ],
  Array [
    Object {
      "baseUrl": "https://example.com/cdn/a/2.2.2/",
      "browser": Object {
        "integrity": "fhgnt543",
        "url": "https://example.com/cdn/a/2.2.2/a.browser.js",
      },
    },
    "a",
  ],
  Array [
    Object {
      "baseUrl": "https://example.com/cdn/b/2.2.2/",
      "browser": Object {
        "integrity": "yhrtrhw3",
        "url": "https://example.com/cdn/b/2.2.2/b.browser.js",
      },
    },
    "b",
  ],
]
`;

exports[`renderInitialStateScriptTag should render initial state inside a script tag body 1`] = `
"<script id=\\"initial-state\\" nonce=\\"19219fb1-8768-4ffa-983c-742b3608589b\\">
	window.__render_mode__ = 'hydrate';
	window.__webpack_public_path__ = \\"/\\";
	window.__holocron_module_bundle_type__ = 'browser';
	window.__CLIENT_HOLOCRON_MODULE_MAP__ = {\\"browser\\":{\\"key\\":\\"123\\",\\"modules\\":{\\"test-root\\":{\\"baseUrl\\":\\"https://example.com/cdn/test-root/2.2.2/\\",\\"browser\\":{\\"url\\":\\"https://example.com/cdn/test-root/2.2.2/test-root.browser.js\\",\\"integrity\\":\\"dhhfsdfwer\\"}},\\"a\\":{\\"baseUrl\\":\\"https://example.com/cdn/a/2.2.2/\\",\\"browser\\":{\\"url\\":\\"https://example.com/cdn/a/2.2.2/a.browser.js\\",\\"integrity\\":\\"fhgnt543\\"}},\\"c\\":{\\"baseUrl\\":\\"https://example.com/cdn/c/2.2.2/\\",\\"browser\\":{\\"url\\":\\"https://example.com/cdn/c/2.2.2/c.browser.js\\",\\"integrity\\":\\"323egdsbf\\"}},\\"b\\":{\\"baseUrl\\":\\"https://example.com/cdn/b/2.2.2/\\",\\"browser\\":{\\"url\\":\\"https://example.com/cdn/b/2.2.2/b.browser.js\\",\\"integrity\\":\\"yhrtrhw3\\"}}}},\\"legacyBrowser\\":{\\"key\\":\\"123\\",\\"modules\\":{\\"test-root\\":{\\"baseUrl\\":\\"https://example.com/cdn/test-root/2.2.2/\\",\\"legacyBrowser\\":{\\"url\\":\\"https://example.com/cdn/test-root/2.2.2/test-root.legacy.browser.js\\",\\"integrity\\":\\"567ty\\"}},\\"a\\":{\\"baseUrl\\":\\"https://example.com/cdn/a/2.2.2/\\",\\"legacyBrowser\\":{\\"url\\":\\"https://example.com/cdn/a/2.2.2/a.legacy.browser.js\\",\\"integrity\\":\\"7567ee\\"}},\\"c\\":{\\"baseUrl\\":\\"https://example.com/cdn/c/2.2.2/\\",\\"legacyBrowser\\":{\\"url\\":\\"https://example.com/cdn/c/2.2.2/c.legacy.browser.js\\",\\"integrity\\":\\"d8vdfdfv\\"}},\\"b\\":{\\"baseUrl\\":\\"https://example.com/cdn/b/2.2.2/\\",\\"legacyBrowser\\":{\\"url\\":\\"https://example.com/cdn/b/2.2.2/b.legacy.browser.js\\",\\"integrity\\":\\"7567ee\\"}}}}};
	window.__INITIAL_STATE__ = {\\"intl\\":{\\"activeLocale\\":\\"en-US\\"},\\"config\\":{\\"rootModuleName\\":\\"test-root\\"},\\"holocron\\":{\\"loaded\\":[\\"test-root\\"]}};
</script>"
`;

exports[`renderScriptTag should render bare bones script tag 1`] = `"<script></script>"`;

exports[`renderScriptTag should render script tag with src 1`] = `"<script src=\\"/app.js\\"></script>"`;

exports[`renderScriptTag should render script tag with src and attrs 1`] = `"<script src=\\"/app.js\\" crossorigin=\\"anonymous\\"></script>"`;

exports[`renderScriptTag should render script tag with src, attrs and body 1`] = `
"<script src=\\"/app.js\\" crossorigin=\\"anonymous\\">
console.log(this);
</script>"
`;

exports[`renderScripts should render all app scripts with bare state as default 1`] = `
"<script id=\\"initial-state\\" nonce=\\"19219fb1-8768-4ffa-983c-742b3608589b\\">
	window.__render_mode__ = 'hydrate';
	window.__webpack_public_path__ = \\"/_/static/\\";
	window.__holocron_module_bundle_type__ = 'browser';
	window.__CLIENT_HOLOCRON_MODULE_MAP__ = {\\"key\\":\\"123\\",\\"modules\\":{\\"test-root\\":{\\"baseUrl\\":\\"https://example.com/cdn/test-root/2.2.2/\\",\\"browser\\":{\\"url\\":\\"https://example.com/cdn/test-root/2.2.2/test-root.browser.js\\",\\"integrity\\":\\"dhhfsdfwer\\"}},\\"a\\":{\\"baseUrl\\":\\"https://example.com/cdn/a/2.2.2/\\",\\"browser\\":{\\"url\\":\\"https://example.com/cdn/a/2.2.2/a.browser.js\\",\\"integrity\\":\\"fhgnt543\\"}},\\"c\\":{\\"baseUrl\\":\\"https://example.com/cdn/c/2.2.2/\\",\\"browser\\":{\\"url\\":\\"https://example.com/cdn/c/2.2.2/c.browser.js\\",\\"integrity\\":\\"323egdsbf\\"}},\\"b\\":{\\"baseUrl\\":\\"https://example.com/cdn/b/2.2.2/\\",\\"browser\\":{\\"url\\":\\"https://example.com/cdn/b/2.2.2/b.browser.js\\",\\"integrity\\":\\"yhrtrhw3\\"}}}};
	window.__INITIAL_STATE__ = \\"{\\\\\\"intl\\\\\\":{\\\\\\"activeLocale\\\\\\":\\\\\\"en-US\\\\\\"},\\\\\\"config\\\\\\":{\\\\\\"rootModuleName\\\\\\":\\\\\\"test-root\\\\\\"},\\\\\\"holocron\\\\\\":{\\\\\\"loaded\\\\\\":[\\\\\\"test-root\\\\\\"]}}\\";
</script>
<script src=\\"/_/static/app~vendors.js\\" integrity=\\"sha256-hQX6eBOiXd4Y1P6lypbnxiSVn5hlngqaQMk4EJpAn44= sha384-dQaNuKynPD1onwRNRxJobv25+Zf3bDF0dgdO9Yl3/a43L548hFiJbOX+BY3xRL0i\\" crossorigin=\\"anonymous\\"></script>
<script src=\\"/_/static/runtime.js\\" integrity=\\"sha256-BS+z4Q61be6nCzJ4wwo7SuTFIMIIGOdCepjFH8MjbSo= sha384-ptJKDvCGQNbG6YFcIGGrTNCGz+/Whizhe0apjJOq3WIsF993jRX5c864OTSaXfEE\\" crossorigin=\\"anonymous\\"></script>
<script src=\\"/_/static/vendors.js\\" integrity=\\"sha256-VaK5IR+K3vP1HaYNTOiTS+6dFA+4cyXrllRsCC8VO+w= sha384-NMQgmNz2kyu39sfgXy88kxQ9/tT6MUNkQNMUYztNScoWLNeTsVrl4Czhw3OyygNc\\" crossorigin=\\"anonymous\\"></script>
<script src=\\"/_/static/i18n/en-US.js\\" crossorigin=\\"anonymous\\"></script>
<script src=\\"https://example.com/cdn/test-root/2.2.2/test-root.browser.js?key=123\\" crossorigin=\\"anonymous\\" integrity=\\"dhhfsdfwer\\"></script>
<script src=\\"/_/static/app.js\\" integrity=\\"sha256-2QXp81GxT/SYSwtjdl48T2cl2fCF6e1n2xoNnhDwpW4= sha384-ZxiHrSuvGSAExxL77bA0MH0Zjx15JUKzaxZqY+04qt1vxGaMamvfn8oXfXshISQk\\" crossorigin=\\"anonymous\\"></script>"
`;

exports[`renderScripts should render scripts without defaults 1`] = `
"<script id=\\"initial-state\\">
	window.__render_mode__ = 'hydrate';
	window.__webpack_public_path__ = ;
	window.__holocron_module_bundle_type__ = 'browser';
	window.__CLIENT_HOLOCRON_MODULE_MAP__ = {};
	window.__INITIAL_STATE__ = \\"{\\\\\\"intl\\\\\\":{\\\\\\"activeLocale\\\\\\":\\\\\\"en-US\\\\\\"},\\\\\\"config\\\\\\":{\\\\\\"rootModuleName\\\\\\":\\\\\\"test-root\\\\\\"},\\\\\\"holocron\\\\\\":{\\\\\\"loaded\\\\\\":[]}}\\";
</script>"
`;

exports[`serializeClientInitialState should fail to serialize state and fallback to defaults 1`] = `"{\\"config\\":{\\"rootModuleName\\":\\"test-root\\"},\\"holocron\\":{\\"loaded\\":[]}}"`;

exports[`serializeClientInitialState should serialize state 1`] = `"{\\"holocron\\":{\\"loaded\\":[]}}"`;
